[SYCL] Add Level-Zero interop with specification of ownership (#3231)
intel/llvm: sycl/doc/extensions/LevelZeroBackend/LevelZeroBackend.md

|-------------|:------------|
|``` make<platform>(ze_driver_handle_t);```|Constructs a SYCL platform instance from a Level-Zero ```ze_driver_handle_t```.|
|``` make<device>(const platform &, ze_device_handle_t);```|Constructs a SYCL device instance from a Level-Zero ```ze_device_handle_t```. The platform argument gives a SYCL platform, encapsulating a Level-Zero driver supporting the passed Level-Zero device.|

> |``` make<context>(const vector_class<device> &, ze_context_handle_t, ownership = transfer);```| Constructs a SYCL context instance from a Level-Zero ```ze_context_handle_t```. The context is created against the devices passed in. There must be at least one device given and all the devices must be from the same SYCL platform and thus from the same Level-Zero driver. The ```ownership``` argument specifies if SYCL RT should take ownership of the passed native handle. The default behavior is to transfer the ownership to SYCL RT. See section 4.4 for details.|

Reviewer: ```suggestion
|``` make<context>(const vector_class<device> &, ze_context_handle_t, ownership = transfer);```| Constructs a SYCL context instance from a Level-Zero ```ze_context_handle_t```. The context is created against the devices passed in. There must be at least one device given and all the devices must be from the same SYCL platform and thus from the same Level-Zero driver. The ```ownership``` argument specifies if the SYCL runtime should take ownership of the passed native handle. The default behavior is to transfer the ownership to the SYCL runtime. See section 4.4 for details.|

```

### 4.4 Level-Zero handles' ownership and thread-safety
        
The Level-Zero runtime doesn't do reference-counting of its objects, so it is crucial to adhere to these
practices of how Level-Zero handles are manged. By default, the ownership us transferred to SYCL RT, but
some interoparability API supports overriding this behavior and keep the ownership in the application.
Use this enumeration for explicit specification of the ownership:
``` C++
enum ownership { transfer, keep };

> ```

Author: OK

                
Whenever the application creates a SYCL object from the corresponding Level-Zero handle via one of the ```make<T>()``` functions,
the SYCL runtime takes ownership of the Level-Zero handle, if no explicit ```ownership::keep``` was specified.
The application must not use the Level-Zero handle after the last host copy of the SYCL object is destroyed (
as described in the core SYCL specification under "Common reference semantics"), and the application must not
destroy the Level-Zero handle itself.
                                                                
#### 4.4.2 SYCL runtime assumes ownership (default)

The application may call the ```get_native<T>()``` member function of a SYCL object to retrieve the underlying Level-Zero handle,
however, the SYCL runtime continues to retain ownership of this handle, unless SYCL object was created with interoperability API that asked
to keep the ownership by the application with ```ownership::keep```. The application must not use this handle after the last host copy of
the SYCL object is destroyed (as described in the core SYCL specification under "Common reference semantics"), and the application must

> not destroy the Level-Zero handle.

Reviewer: I would move this section below "Application keeps ownership (explicit)".  I would also rename the section title to match what is really being described here.  So, the outline would become:

* 4.4.1 SYCL runtime takes ownership (default)
* 4.4.2 Application keeps ownership (explicit)
* 4.4.3 Obtaining native handle does not change ownership

I would also change the text of this section to focus on the fact that the ownership is not changed when calling `get_native()`.  For example:

```
The application may call the ```get_native<T>()``` member function of a SYCL object to retrieve the underlying Level-Zero handle.
Doing so does not change the ownership of the the Level-Zero handle.  Therefore, the application may not use this
handle after the last host copy of the SYCL object is destroyed (as described in the core SYCL specification under
"Common reference semantics") unless the SYCL object was created by the application with ```ownership::keep```.

Author: Thanks, I think this is a good suggestion. Will follow

#### 4.4.3 Application keeps ownership (explicit)

If SYCL object is created with an interoperability API explicitly asking to keep the native handle ownership in the application with

> ```ownership::keep``` then SYCL RT does not take the ownership and will not destroy the Level-Zero handle at the destruction of the SYCL object.

Reviewer: ```suggestion
```ownership::keep``` then the SYCL runtime does not take the ownership and will not destroy the Level-Zero handle at the destruction of the SYCL object.
```

> Application is responsible for destroying the native handle when it no longer needs it, but not earlier than the SYCL object created with that
> handle is EOL.

Reviewer: ```suggestion
The application is responsible for destroying the native handle when it no longer needs it, but it must not destroy the
handle before the last host copy of the SYCL object is destroyed (as described in the core SYCL specification under
"Common reference semantics").
```
